File: HW 5 (kotlin, room)/contacts/src/main/java/com/example/contacts/EditContactActivity.java
package com.example.contacts;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;


public class EditContactActivity extends AppCompatActivity {
    enum CONTACT_MODIFICATION{
        REMOVE,
        EDIT
    }

    private EditText nameField;
    private EditText contactField;
    private Button removeButton;
    private Toolbar toolbar;
    private ActionBar actionBar;

    private Contact receivedContact;
    private CONTACT_MODIFICATION modification;

    public static final String KEY_CHANGED_CONTACT = "KEY CHANGED CONTACT";
    public static final String REMOVE_OR_EDIT_CONTACT = "REMOVE OR EDIT CONTACT";
    public static final String KEY_CONTACT_FOR_EDITING = "CONTACT FOR EDITING";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.edit_contact);

        nameField = findViewById(R.id.personName_field);
        contactField = findViewById(R.id.personContact_field);
        removeButton = findViewById(R.id.button_remove);
        toolbar = findViewById(R.id.toolbarEditActivity);
        setSupportActionBar(toolbar);

        removeButton.setOnClickListener(removeButtonListener);

        actionBar = getSupportActionBar();
        if(actionBar != null){
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        receiveContact();
        setFields(receivedContact);
    }

    public static Intent newIntent(Context context, Contact contact){
        Intent intent = new Intent(context, EditContactActivity.class);
        intent.putExtra(KEY_CONTACT_FOR_EDITING, contact);
        return intent;
    }

    private void receiveContact(){
        Intent intent = getIntent();
        receivedContact = (Contact) intent.getExtras().getSerializable(KEY_CONTACT_FOR_EDITING);
    }

    private void setFields(Contact contact){
        nameField.setText(contact.getPersonName());
        contactField.setText(contact.getContactDetails());
    }

    private Contact modifyContact(Contact contact){
        contact.setPersonName(nameField.getText().toString());
        contact.setContactDetails(contactField.getText().toString());
        return contact;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            modification = CONTACT_MODIFICATION.EDIT;
            modifyContact(receivedContact);
            sendChangesBack();
        }
        return super.onOptionsItemSelected(item);
    }

    private View.OnClickListener removeButtonListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            modification = CONTACT_MODIFICATION.REMOVE;
            modifyContact(receivedContact);
            sendChangesBack();
        }
    };

    private void sendChangesBack(){
        Intent resultIntent = new Intent();
        resultIntent.putExtra(KEY_CHANGED_CONTACT, receivedContact);
        resultIntent.putExtra(REMOVE_OR_EDIT_CONTACT, modification);
        setResult(Activity.RESULT_OK, resultIntent);
        finish();
    }
}

File: HW 5/kotlin_sql/src/main/java/com/example/kotlin_sql/EditContactActivity.kt
package com.example.kotlin_sql

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import com.example.kotlin_sql.CONTACT_MODIFICATION.EDIT
import com.example.kotlin_sql.CONTACT_MODIFICATION.REMOVE
import kotlinx.android.synthetic.main.edit_contact.*

const val KEY_CHANGED_CONTACT: String = "KEY CHANGED CONTACT"
const val REMOVE_OR_EDIT_CONTACT: String = "REMOVE OR EDIT CONTACT"
private const val KEY_CONTACT_FOR_EDITING: String = "CONTACT FOR EDITING"

class EditContactActivity : AppCompatActivity() {

    private var actionBar: ActionBar? = null
    private var receivedContact: Contact? = null
    private var modification: CONTACT_MODIFICATION? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.edit_contact)

        setSupportActionBar(toolbarEditActivity)

        removeButton?.setOnClickListener(removeButtonListener)

        actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)


        receiveContact()
        receivedContact?.let { setFields(it) }
    }

    fun newIntent(context: Context, contact: Contact): Intent {
        val intent = Intent(context, EditContactActivity::class.java)
        intent.putExtra(KEY_CONTACT_FOR_EDITING, contact)
        return intent
    }

    private fun receiveContact() {
        val intent = intent
        receivedContact = intent.extras?.getSerializable(KEY_CONTACT_FOR_EDITING) as Contact
    }

    private fun setFields(contact: Contact) {
        personNameField.text = contact.personName as Editable
        personContactField.text = contact.contactDetails as Editable
    }

    private fun modifyContact(contact: Contact): Contact {
        contact.personName = personNameField?.text.toString()
        contact.contactDetails = personContactField?.text.toString()
        return contact
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            modification = EDIT
            modifyContact(receivedContact!!)
            sendChangesBack()
        }
        return super.onOptionsItemSelected(item)
    }

    private val removeButtonListener = View.OnClickListener {
        modification = REMOVE
        modifyContact(receivedContact!!)
        sendChangesBack()
    }

    private fun sendChangesBack() {
        val resultIntent = Intent()
        resultIntent.putExtra(KEY_CHANGED_CONTACT, receivedContact)
        resultIntent.putExtra(REMOVE_OR_EDIT_CONTACT, modification)
        setResult(RESULT_OK, resultIntent)
        finish()
    }
}